### PR TITLE
Prepare for Auto-flush, introduce listeners to carry state with callback API

### PR DESCRIPTION
**Pull request by guyboertje**
NOTES TO REVIEWERS:
- Both the IdentityMapCodec and the Multiline have been changed.
- Both are backward compatible and retain the block based API.
- New `accept` method that takes a listener as an argument.
- For these two classes, Listeners must implement `accept()` and `process_event(event)`.
- Listeners usually hold long term upstream state (context) and transient data for downstream to process.
- The listener based tests are a WIP and will be tidied up.
- A lot more comments need to be added.
- Some classes will need to be in their own files but are included for readability.
- These changes are a prelude to a better component based dataflow channel in each input.
- Auto-flush is implemented.

### CODE REVIEW

File: spec/supports/helpers.rb [hunk 1]
```diff
@@ -10,3 +10,71 @@ def decode_events
...
+
+  # dup and add state for downstream
+  def dup_adding_state(line)
+    self.class.new(queue, codec, path).add_state(line)
+  end
+end
+
+class LineErrorListener < LineListener
+  def process_event(event)
+    raise StandardError.new("OMG, Darleks!")
```
Comment (ph) [inline spec/supports/helpers.rb]: Daleks! Exterminate.

File: lib/logstash/codecs/multiline.rb [hunk 2]
```diff
@@ -126,7 +127,68 @@ class LogStash::Codecs::Multiline < LogStash::Codecs::Base
...
+    end
+
+    def stopped?
+      true
+    end
+
+    def method_missing(*args, &block)
+      self
+    end
+  end
```
Comment (ph) [inline lib/logstash/codecs/multiline.rb]: Have you considered defining all the possible methods instead of using `method_missing`?

Comment (ph) [inline lib/logstash/codecs/multiline.rb]: Also I have made this [matcher](https://github.com/elastic/logstash/blob/feature/metrics/logstash-core/spec/support/matchers.rb#L14) in my metric branch to test the null object with great result so far.

File: lib/logstash/codecs/multiline.rb [hunk 3]
```diff
@@ -126,7 +127,68 @@ class LogStash::Codecs::Multiline < LogStash::Codecs::Base
...
+        @task = Concurrent::ScheduledTask.execute(@interval, &@run_block)
+      # else the task is executing
+      end
+      self
+    end
+
+    def finished?
+      return true if @task.nil?
+      @task.fulfilled?
+    end
```
Comment (ph) [inline lib/logstash/codecs/multiline.rb]: They have a sweet api.

File: lib/logstash/codecs/multiline.rb [hunk 4]
```diff
@@ -126,7 +127,68 @@ class LogStash::Codecs::Multiline < LogStash::Codecs::Base
...
+
+    def stopped?
+      @stopped
+    end
+
+    def stop
+      @stopped = true
+      if pending?
+        @task.cancel
+      end
```
Comment (ph) [inline lib/logstash/codecs/multiline.rb]: one liner, maybe?

File: lib/logstash/codecs/multiline.rb [hunk 5]
```diff
@@ -175,23 +249,41 @@ def decode(text, &block)
...
     end
   end
 
   def merge_events
     event = LogStash::Event.new(LogStash::Event::TIMESTAMP => @time, "message" => @buffer.join(NL))
     event.tag @multiline_tag if @multiline_tag && @buffer.size > 1
-    event.tag "multiline_codec_max_bytes_reached" if over_maximun_bytes?
-    event.tag "multiline_codec_max_lines_reached" if over_maximun_lines?
+    event.tag "multiline_codec_max_bytes_reached" if over_maximum_bytes?
+    event.tag "multiline_codec_max_lines_reached" if over_maximum_lines?
```
Comment (ph) [inline lib/logstash/codecs/multiline.rb:231]: thanks for correcting the typo

File: lib/logstash/codecs/multiline.rb (hunk 2 above)
Comment (guyboertje) [inline lib/logstash/codecs/multiline.rb]: Whether or not the `AutoFlushUnset` is used at all is dependent on whether the config option should have a default.  If we say it should then `AutoFlushUnset` is never used.

File: lib/logstash/codecs/multiline.rb (hunk 5 above)
Comment (guyboertje) [inline lib/logstash/codecs/multiline.rb:231]: sure, you're welcome.

File: lib/logstash/codecs/multiline.rb (hunk 4 above)
Comment (guyboertje) [inline lib/logstash/codecs/multiline.rb]: OK

File: spec/codecs/multiline_spec.rb [hunk 6]
```diff
@@ -218,4 +214,110 @@
...
+        expect(events.size).to eq(1)
+        line_producer.call("de.log")
+        sleep(auto_flush_interval + 0.1)
+        expect(events.size).to eq(3)
+        expect(events[1]["path"]).to eq("fr.log")
+        expect(events[1]["message"]).to eq("Salut le Monde\n deuxième ligne\n troisième ligne")
+        expect(events[2]["path"]).to eq("de.log")
+        expect(events[2]["message"]).to eq("Hallo Welt")
+        expect(codec.buffer_size).to eq(0)
+      end
```
Comment (ph) [inline spec/codecs/multiline_spec.rb:295]: That a lot of assertions in this method, its is possible to break it down?

File: spec/codecs/multiline_spec.rb [hunk 7]
```diff
@@ -218,4 +214,110 @@
...
+        expect(events[1]["path"]).to eq("de.log")
+        expect(events[1]["message"]).to eq("Hallo Welt")
+        expect(codec.buffer_size).to eq(0)
+        line_producer.call("fr.log")
+        sleep(auto_flush_interval + 0.1)
+        expect(events.size).to eq(3)
+        expect(events[2]["path"]).to eq("fr.log")
+        expect(events[2]["message"]).to eq("Salut le Monde++\ndeuxième ligne++\ntroisième ligne")
+        expect(codec.buffer_size).to eq(0)
+      end
```
Comment (ph) [inline spec/codecs/multiline_spec.rb:324]: lot of assertions here too.

File: spec/supports/helpers.rb [hunk 8]
```diff
@@ -10,3 +10,71 @@ def decode_events
...
+
+  def trace_for(symbol)
+    params = @tracer.assoc(symbol)
+    params.nil? ? false : params.last
+  end
+
+  def clear()
+    @tracer.clear()
+  end
+end
```
Comment (ph) [inline spec/supports/helpers.rb:78]: oh love that!

File: spec/codecs/multiline_spec.rb (hunk 6 above)
Comment (ph) [inline spec/codecs/multiline_spec.rb:295]: A customs matcher could do the `event` assert and would remove some lines?

**Comment by ph** [comment]
Added minor comments, address them and we can merge.

Also I fixed the issue with `concurrent-ruby` on the jenkins server you can see a build of this PR at http://build-eu-00.elastic.co/job/logstash-plugin-codec-multiline_PR/10/console

File: lib/logstash/codecs/multiline.rb (hunk 2 above)
Comment (ph) [inline lib/logstash/codecs/multiline.rb]: but, if its shouldn't It use `autoflushnet`, so we should unit test it no?

File: spec/codecs/multiline_spec.rb [hunk 9]
```diff
@@ -218,4 +214,121 @@
...
+        line_producer.call("fr.log")
+        sleep(auto_flush_interval - 0.3) #next line(s) come before auto-flush i.e. buffered
+        expect(codec.buffer_size).to eq(3)
+        expect(events.size).to eq(1)
+
+        line_producer.call("de.log")
+        sleep(auto_flush_interval + 0.1)
+        expect(events[1]).to match_path_and_line("fr.log", lines["fr.log"])
+        expect(events[2]).to match_path_and_line("de.log", lines["de.log"])
+        expect(codec).to have_an_empty_buffer
```
Comment (guyboertje) [inline spec/codecs/multiline_spec.rb]: @ph - please tell me which of the two styles you prefer lines 272-286 or lines 310-312

**Comment by guyboertje** [comment]
Oops - changed the evict method.  will fix test.

**Comment by guyboertje** [comment]
specs pass locally

**Comment by guyboertje** [comment]
again specs pass locally after spec helper re-org.

**Comment by ph** [comment]
Make sure you add the changes to the changelod.md and bump the version so I can review those and make sure we squash the PR.

but the code lgtm and the test pass.

**Comment by guyboertje** [comment]
OK.

**Comment by ph** [comment]
LGTM

**Pull request merged**